### PR TITLE
Added field to SentTextMessage, extended FFI

### PR DESCRIPTION
**Pull request by StriderDM**
Added boolean field to SentTextMessage.
Extended FFI to return additional fields for ReceivedTextMessage.
Extended FFI for SentTextMessage.

## Description
Added boolean field isread to SentTextMessage in Model, Schema and Migration.
Added Test to SentTextMessage for Mark_Sent_Message_Opened method.
Implemented Ord and PartialOrd for SentTextMessage.
Extended FFI to return additional fields for ReceivedTextMessage.
Extended FFI for SentTextMessage.

## Motivation and Context
Useful field to include, differentiates between a message being delivered and the message being processed.
FFI needed to be extended to account for Sent messages, Received messages needed to be able to return additional fields.

## How Has This Been Tested?
cargo test --all

## Types of changes
* [ ] Bug fix (non-breaking change which fixes an issue)
* [ ] New feature (non-breaking change which adds functionality)
* [x] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [x] New Tests
* [ ] Documentation

## Checklist:

* [x] I'm merging against the `development` branch
* [x] I ran `cargo-fmt --all` before pushing
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [ ] I have added tests to cover my changes.

### CODE REVIEW

File: base_layer/wallet/migrations/2019-06-26-130555_initial/up.sql [hunk 1]
```diff
@@ -5,6 +5,7 @@ CREATE TABLE sent_messages (
    message TEXT  NOT NULL,
    timestamp DATETIME NOT NULL,
    acknowledged INTEGER NOT NULL DEFAULT 0,
+   isread INTEGER NOT NULL DEFAULT 0,
```
Comment (sdbondi) [inline base_layer/wallet/migrations/2019-06-26-130555_initial/up.sql]: `snake_case` is preferred. Please change to `is_read`

File: base_layer/wallet/src/schema.rs [hunk 2]
```diff
@@ -24,6 +24,7 @@ table! {
         message -> Text,
         timestamp -> Timestamp,
         acknowledged -> Integer,
+        isread -> Integer,
```
Comment (sdbondi) [inline base_layer/wallet/src/schema.rs]: `is_read` 

Shouldn't this be a Bool?

File: base_layer/wallet/src/text_message_service/model.rs [hunk 3]
```diff
@@ -62,14 +62,15 @@ pub fn generate_id<D: Digest>(
...
-#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
 pub struct SentTextMessage {
     pub id: Vec<u8>,
     pub source_pub_key: CommsPublicKey,
     pub dest_pub_key: CommsPublicKey,
     pub message: String,
     pub timestamp: NaiveDateTime,
     pub acknowledged: bool,
+    pub isread: bool,
```
Comment (sdbondi) [inline base_layer/wallet/src/text_message_service/model.rs]: Replace `isread` with `is_read` in this file

File: ffi/src/lib.rs [hunk 4]
```diff
@@ -42,17 +42,24 @@ use tari_crypto::keys::PublicKey;
 use tari_p2p::{initialization::CommsConfig, sync_services::ServiceError};
 use tari_utilities::hex::Hex;
 use tari_wallet::{text_message_service::Contact, wallet::WalletConfig};
+// use std::os::raw::c_void;
```
Comment (sdbondi) [inline ffi/src/lib.rs]: Remove commented out code

File: ffi/src/lib.rs [hunk 5]
```diff
@@ -78,6 +85,68 @@ pub struct Peers {
...
+/// Sets the isread of a SentText Message, returns if successful or not
+#[no_mangle]
+pub unsafe extern "C" fn senttextmessage_set_opened(o: *mut SentTextMessage, w: *mut Wallet) -> bool {
+    if !w.is_null() {
+        if !o.is_null() {
+            if !(*o).acknowledged {
+                return false; // can't be opened before being delivered
+            }
+
+            if (*o).isread {
```
Comment (sdbondi) [inline ffi/src/lib.rs]: Replace `isread` with `is_read` in this file

File: ffi/src/lib.rs [hunk 6]
```diff
@@ -303,6 +447,49 @@ pub unsafe extern "C" fn wallet_get_receivedmessages(w: *mut Wallet) -> *mut Rec
...
+        .text_message_service
+        .get_text_messages()
+        .expect("Error retrieving text messages from TMS")
+        .sent_messages;
+
+    rx_messages.sort();
+
+    let mut messages = SentMessages(Vec::new());
+
+    for i in 0..rx_messages.len() {
```
Comment (sdbondi) [inline ffi/src/lib.rs]: Try this:

```rust
let messages = rx_messages.into_iter()
  // Filter out any messages if the pub_key is not in our contact list
  .filter(|msg| contacts.iter().any(|c| c.pub_key == msg.source_pub_key))
  .collect::<Vec<_>>();
let messages = SentMessages::new(messages);
```

Add a `new` function to SentMessages

File: base_layer/wallet/src/schema.rs (hunk 2 above)
Comment (StriderDM) [reply base_layer/wallet/src/schema.rs]: There is no default boolean type in SQLLite:

https://www.sqlite.org/datatype3.html

File: base_layer/wallet/src/text_message_service/model.rs [hunk 7]
```diff
@@ -670,6 +709,11 @@ mod test {
         let find3 = SentTextMessage::find(&sent_msg1.id, &conn).unwrap();
         assert!(find3.acknowledged);
 
+        assert!(SentTextMessage::mark_sent_message_opened(vec![2u8; 32], &conn).is_err());
+        SentTextMessage::mark_sent_message_opened(sent_msg1.clone().id, &conn).unwrap();
```
Comment (neonknight64) [inline base_layer/wallet/src/text_message_service/model.rs]:
```suggestion
        SentTextMessage::mark_sent_message_opened(sent_msg1.id.clone(), &conn).unwrap();
```

File: ffi/src/lib.rs [hunk 8]
```diff
@@ -78,6 +84,68 @@ pub struct Peers {
     peers: Vec<ConfigPeer>,
 }
 
+/// Sets the is_read of a SentText Message, returns if successful or not
+#[no_mangle]
+pub unsafe extern "C" fn senttextmessage_set_opened(o: *mut SentTextMessage, w: *mut Wallet) -> bool {
+    if !w.is_null() {
+        if !o.is_null() {
```
Comment (neonknight64) [inline ffi/src/lib.rs]: You can probably combine ln90 and ln91 into a single if with an 'and'

File: ffi/src/lib.rs [hunk 9]
```diff
@@ -78,6 +84,68 @@ pub struct Peers {
...
+            }
+
+            if (*o).is_read {
+                return true; // already read, no change needed
+            }
+
+            let rx: Option<Vec<SentTextMessage>> =
+                Some((*w).text_message_service.get_text_messages().unwrap().sent_messages);
+
+            if !rx.is_some() {
```
Comment (neonknight64) [inline ffi/src/lib.rs]: In ln101, rx will always be `Some`, it might not have to be an`Option`

File: ffi/src/lib.rs [hunk 10]
```diff
@@ -78,6 +84,68 @@ pub struct Peers {
...
+        }
+    }
+    return false;
+}
+
+/// Returns the is_read of a SentText Message as bool
+#[no_mangle]
+pub unsafe extern "C" fn senttextmessage_get_opened(o: *const SentTextMessage) -> bool {
+    if !o.is_null() {
+        (*o).is_read;
```
Comment (neonknight64) [inline ffi/src/lib.rs]:
```suggestion
        return (*o).is_read;
```

File: ffi/src/lib.rs [hunk 11]
```diff
@@ -78,6 +84,68 @@ pub struct Peers {
...
+    }
+    false
+}
+
+/// Returns the acknowledged of a SentText Message as bool
+#[no_mangle]
+pub unsafe extern "C" fn senttextmessage_get_acknowledged(o: *const SentTextMessage) -> bool {
+    if !o.is_null() {
+        (*o).acknowledged;
```
Comment (neonknight64) [inline ffi/src/lib.rs]:
```suggestion
        return (*o).acknowledged;
```

**Pull request merged**